### PR TITLE
python312Packages.{ax,botorch}: relax version requirements

### PR DESCRIPTION
**Pull request by veprbl**
## Description of changes

This is needed to make incremental package updates possible.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by veprbl** [comment]
@GrahamcOfBorg build python311Packages.ax python311Packages.botorch python312Packages.ax python312Packages.botorch

**Comment by veprbl** [comment]
@GrahamcOfBorg eval

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 328732` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.ax</li>
    <li>python311Packages.ax.dist</li>
    <li>python312Packages.optuna</li>
    <li>python312Packages.optuna.dist</li>
    <li>python312Packages.pyannote-pipeline</li>
    <li>python312Packages.pyannote-pipeline.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>python311Packages.botorch</li>
    <li>python311Packages.botorch.dist</li>
    <li>python311Packages.optuna</li>
    <li>python311Packages.optuna.dist</li>
    <li>python311Packages.pyannote-pipeline</li>
    <li>python311Packages.pyannote-pipeline.dist</li>
    <li>python312Packages.ax</li>
    <li>python312Packages.ax.dist</li>
    <li>python312Packages.botorch</li>
    <li>python312Packages.botorch.dist</li>
  </ul>
</details>

**Comment by corngood** [comment]
Result of `nixpkgs-review pr 328732` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>python311Packages.ax</li>
    <li>python311Packages.ax.dist</li>
    <li>python311Packages.botorch</li>
    <li>python311Packages.botorch.dist</li>
    <li>python311Packages.optuna</li>
    <li>python311Packages.optuna.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pyannote-pipeline</li>
    <li>python311Packages.pyannote-pipeline.dist</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python312Packages.ax</li>
    <li>python312Packages.ax.dist</li>
    <li>python312Packages.botorch</li>
    <li>python312Packages.botorch.dist</li>
    <li>python312Packages.optuna</li>
    <li>python312Packages.optuna.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-pipeline</li>
    <li>python312Packages.pyannote-pipeline.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by corngood** [comment]
Result of `nixpkgs-review pr 328732` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.ax</li>
    <li>python311Packages.ax.dist</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>python311Packages.botorch</li>
    <li>python311Packages.botorch.dist</li>
    <li>python311Packages.optuna</li>
    <li>python311Packages.optuna.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pyannote-pipeline</li>
    <li>python311Packages.pyannote-pipeline.dist</li>
    <li>python312Packages.ax</li>
    <li>python312Packages.ax.dist</li>
    <li>python312Packages.botorch</li>
    <li>python312Packages.botorch.dist</li>
    <li>python312Packages.optuna</li>
    <li>python312Packages.optuna.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-pipeline</li>
    <li>python312Packages.pyannote-pipeline.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by afh** [comment]
ℹ️ botorch [version 0.11.3](https://github.com/pytorch/botorch/releases/tag/v0.11.3) was released recently and has bumped the gpytorch requirement to 0.12.

File: pkgs/development/python-modules/botorch/default.nix [hunk 1]
```diff
@@ -26,6 +26,11 @@ buildPythonPackage rec {
     hash = "sha256-YX/G46U09y/VZuWZhKY8zU0Y+bf0NKumzSGYUWvrq/0=";
   };
 
+  postPatch = ''
+    substituteInPlace requirements.txt \
+      --replace-warn 'gpytorch==' 'gpytorch>='
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/botorch/default.nix:31]: Does `relaxPythonDeps` not handle this?

**Comment by afh** [comment]
Great suggestion, @SomeoneSerge, and thanks for merging, @mweinelt, very helpful 💯

**Comment by SomeoneSerge** [comment]
Botorch still broken on master, I'm not sure `pythonRelaxDeps` has actually worked

**Comment by SomeoneSerge** [comment]
Ahh:) `dependenciess` instead of `dependencies` xD

**Pull request merged**